upstairs/src/guest.rs

```diff
@@ -127,20 +90,43 @@ impl GtoS {
          * given up because an IO took too long, or other possible
          * guest side reasons.
          */
-        if let Some(res) = self.res {
-            match result {
-                Ok(_) => match guest_buffer {
-                    Some(guest_buffer) => res.send_ok_with_buffer(guest_buffer),
-                    None => res.send_ok(),
-                },
-
-                Err(e) => match guest_buffer {
-                    Some(guest_buffer) => {
-                        res.send_err_with_buffer(guest_buffer, e)
+        match self.res {
+            Some(GuestBlockRes::Read(mut buffer, res)) => {
+                if let Some(downstairs_responses) = downstairs_responses {
+                    let mut offset = 0;
+
+                    // XXX don't do if result.is_err()?
+                    for response in &downstairs_responses {
+                        // Copy over into guest memory.
+                        {
+                            let _ignored =
+                                span!(Level::TRACE, "copy to guest buffer")
+                                    .entered();
+
+                            buffer.write_read_response(offset, response);
+                            offset += response.data.len();
+                        }
                     }
-                    None => res.send_err(e),
-                },
+                } else {
+                    // Should this panic?  If the caller is requesting a
```

Comment:
This is a pre-existing comment, so I didn't think too hard about it:
https://github.com/oxidecomputer/crucible/blob/89662ead77a0ad4c278677253e2d46ffe8f94830/upstairs/src/guest.rs#L106-L114

Reply:
It may have been out of date before this change came along.
In this path, the buffer has to exist, so the comment can't be
true with regards to that.  I also think even before that, there
were problems here.  I can come in later an fix this to reflect reality.

I think the `else` here is only possible if we had no downstairs_responses
then the IO has failed in some way where we did not get a response
at all. 

I suspect if we failed all three reads we could not have any
downstairs_responses.  In that case we could verify that the result
was an error.  I'll ponder it a bit more to see if there is any condition
where we are not returning an error, but don't have any downstairs_response.